### PR TITLE
uss using zftp API

### PR DESCRIPTION
**Pull request by tiantn**
Signed-off-by: tian na <tiantn@cn.ibm.com>

## Proposed changes

Updated zowe-explorer-ftp-extension under packages to use new zftp APIs.

## Release Notes

It depends on the PR https://github.com/zowe/zowe-cli-ftp-plugin/pull/67

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts [hunk 1]
```diff
@@ -183,10 +204,14 @@ export class FtpUssApi implements ZoweExplorerApi.IUss {
...
         if (connection && connection.client) {
             if (type === "directory") {
-                await connection.client.mkdir(ussPath);
-            } else if (type === "file") {
+                await UssUtils.makeDirectory(connection, ussPath);
+            } else if (type === "File" || type === "file") {
                 const content = Buffer.from(CoreUtils.addCarriageReturns(""));
-                await connection.uploadDataset(content, ussPath, "ascii");
+                const transferOptions = {
+                    transferType: "ascii",
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts:211]: Can use `TRANSFER_TYPE_ASCII` here

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts [hunk 2]
```diff
@@ -183,10 +204,14 @@ export class FtpUssApi implements ZoweExplorerApi.IUss {
         const connection = await this.ftpClient(this.checkedProfile());
         if (connection && connection.client) {
             if (type === "directory") {
```
Comment (std4lqi) [inline packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts:206]: I suppose it's from Zowe Explorer. If type could be `File` or `file`, how about `directory`?

**Review by std4lqi — APPROVED** [review]

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts (hunk 1 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts:211]: I will change it to TRANSFER_TYPE_ASCII. Thanks!

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts (hunk 2 above)
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts:206]: There are two places to create file in vscode, one the type is 'file', one is 'File'. During my tests I found this. No findings on directory. Thank you!

**Review by lauren-li — CHANGES_REQUESTED** [review]
Thanks @tiantn for this update! Overall, it is looking good. I made some requests to update the FTP CLI plugin dependency version and the import statements.

~
~~I am having some trouble getting a USS file rename to work in Zowe Explorer with this PR (zftp profile only). However, I also get an error (but rc 129) when trying to do a USS file rename in the CLI with the Zowe FTP CLI plugin built from @std4lqi's PR (https://github.com/zowe/zowe-cli-ftp-plugin/pull/67), so I will also post my finding there, as it may be an issue on that side, instead. Renaming a USS folder seems to work well.~~

~~Here is the error I see in Zowe Explorer when I try to rename a USS file with my zftp profile:~~

Screen Shot 2020-12-17 at 1 50 02 AM

EDIT: Actually, I am not able to recreate the above error reliably. I will update if I see it again (or create a new issue if I find it is something out-of-scope for this PR).

File: packages/zowe-explorer-ftp-extension/package.json [hunk 3]
```diff
@@ -37,7 +37,7 @@
     "vscode": "^1.43.0"
   },
   "dependencies": {
-    "@zowe/zowe-explorer-api": "1.10.1",
+    "@zowe/zowe-explorer-api": "1.11.0",
     "@zowe/zos-ftp-for-zowe-cli": "1.2.0",
```
[the remark this answers is not in the document]
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/package.json]: Thank you, Lauren. I changed the version to 1.3.0.

File: packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts [hunk 4]
```diff
@@ -17,10 +17,15 @@ import * as zowe from "@zowe/cli";
...
-import { StreamUtils } from "@zowe/zos-ftp-for-zowe-cli/lib/api/StreamUtils";
-import { CoreUtils } from "@zowe/zos-ftp-for-zowe-cli/lib/api/CoreUtils";
+import {
+    CoreUtils,
+    UssUtils,
+    TRANSFER_TYPE_ASCII,
+    TRANSFER_TYPE_BINARY,
+    IZosFTPProfile,
+    FTPConfig
+} from "@zowe/zos-ftp-for-zowe-cli/lib/api";
```
[the remark this answers is not in the document]
Comment (tiantn) [reply packages/zowe-explorer-ftp-extension/src/ZoweExplorerFtpApi.ts]: I removed the /lib/api. Thank you for your review.

**Review by zFernand0 — APPROVED** [review]
LGTM

**Review by lauren-li — APPROVED** [review]
I think this looks good now! All the Zowe Explorer USS functionalities I tested with my zftp profile seem to be working well.

There are some line formattings that Prettier is complaining about when running `yarn run vscode:prepublish`, but @zFernand0 has informed me that we can take care of these and the `package.json` merge conflict on our end while merging in this PR.

Thanks @tiantn for your hard work on this! Also, thank you for providing the list of zftp functionalities you manually tested. I followed this list, as well, and am pasting it below for completeness:

- List uss files and directories
- view file in text/binary mode
- Edit file and save changed
- Create a new directory/new file
- Upload file
- Rename file/directory
- Delete file/directory
- Pull from mainframe
- Save file with conflict (Etag test)

**Comment by zFernand0** [comment]
macos, node10 failed due to an annoying socket timeout.
I'm fine with merging this one : )

**Pull request merged**